src/System.Net.Http/tests/StressTests/HttpStress/StressClient.cs

```diff
@@ -15,174 +18,365 @@ namespace HttpStress
-                                    Console.WriteLine("~~ Reuse address failures: " + reuseAddressFailure.ToString("N0") + "~~");
-                                    Console.ResetColor();
-                                }
-
-                                for (int i = 0; i < clientOperations.Length; i++)
-                                {
-                                    Console.ForegroundColor = ConsoleColor.Cyan;
-                                    Console.Write("\t" + clientOperations[i].Item1.PadRight(30));
-                                    Console.ResetColor();
-                                    Console.ForegroundColor = ConsoleColor.Green;
-                                    Console.Write("Success: ");
-                                    Console.Write(success[i].ToString("N0"));
-                                    Console.ResetColor();
-                                    Console.ForegroundColor = ConsoleColor.Yellow;
-                                    Console.Write("\tCanceled: ");
-                                    Console.Write(cancel[i].ToString("N0"));
-                                    Console.ResetColor();
-                                    Console.ForegroundColor = ConsoleColor.DarkRed;
-                                    Console.Write("\tFail: ");
-                                    Console.ResetColor();
-                                    Console.WriteLine(fail[i].ToString("N0"));
-                                }
-                                Console.WriteLine();
-                            }
-                        }
-                    })
-                    { IsBackground = true }.Start();
+                        current.Failures.TryGetValue(operationIndex, out int failureCount);
+                        return new StressFailureType(current.ErrorText, current.Failures.SetItem(operationIndex, failureCount + 1));
+                    }
 
-                    // Request context factory to be applied in the context of a single worker
-                    Func<RequestContext> CreateRequestContextFactory(int taskNum)
+                    (Type exception, string errorMessage) ClassifyFailure(Exception exn)
                     {
-                        // Creates a System.Random instance that is specific to the current client job
-                        // Generated using the global seed and the task index
-                        Random CreateRandomInstance()
-                        {
-                            // deterministic hashing copied from System.Runtime.Hashing
-                            int Combine(int h1, int h2)
-                            {
-                                uint rol5 = ((uint)h1 << 5) | ((uint)h1 >> 27);
-                                return ((int)rol5 + h1) ^ h2;
-                            }
-
-                            return new Random(Seed: Combine(taskNum, randomSeed));
-                        }
+                        Exception inner = exn;
```

Comment:
Why are we only looking at the innermost exception here?  I'm a little worried that this is going to group together exceptions that aren't really the same. 

For example, when an HTTP2 connection fails, we will store _abortException and use it as the inner exception for other exceptions, but it may get surfaced to the user via multiple different paths. Seems like it's useful to know about these different paths.

Reply:
It's a very rough grouping logic, almost intentionally. I just wanted a way to view all the errors that occured after a long stress run, without having to audit all the logs. The individual traces are still there, but this final report is mostly useful for tuning a test run towards inducing specific types of failures in a shorter timespan.

What alternative do you propose? I was thinking about possibly grouping by list of nested exception types, but couldn't think of a quick way to do it in C#? Do we have an list types that implement structural equality OOTB?

Reply:
Yeah, I realize there is probably no ideal way to do this, or at least not without making it really complicated. My concern here is that once you do grouping, it's easy to forget what might be getting lost in the grouping.

From my point of view, the most useful grouping is by callsite where the exception was thrown. Even that could hide some different paths for how the exception is surfaced, though. Better would be full call stack up to where the code enters System.Net.Http, but that's probably a pain to do.

FYI, Linq has a SequenceEquals operator you can use for sequence equality.

Reply:
Another alternative would be to just list all the failures individually in the report. That's simple and nothing will get lost. It's verbose, but maybe that's better than the alternative.

Reply:
I refined the classification logic so that exceptions are distinguished by nested exceptions, each identified by exception type, exception message and first stackframe.

cf https://github.com/dotnet/corefx/pull/39928/commits/aab093798a2bafea26b738fc898fd2e9c51639d6